File: main/chat_client.py
import asyncio
import websockets
import json
import logging
from datetime import datetime
import urllib.request

# 로깅 설정
logging.basicConfig(level=logging.WARNING, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

class ChzzkChatClient:
    """치지직 채팅 클라이언트"""
    
    def __init__(self, channel_id: str):
        self.channel_id = channel_id
        self.chat_channel_id = None
        self.access_token = None
        self.websocket = None
        self.is_connected = False
        
        # 웹소켓 엔드포인트들
        self.endpoints = [
            f"wss://kr-ss{i}.chat.naver.com/chat" for i in range(1, 6)
        ]
    
    def get_chat_channel_id_sync(self):
        """채팅 채널 ID 조회 (동기 방식) - Old version과 동일한 검증된 로직"""
        # 1단계: 채널 정보 조회로 방송 상태 확인
        channel_info_url = f"https://api.chzzk.naver.com/service/v1/channels/{self.channel_id}"
        
        try:
            req = urllib.request.Request(channel_info_url)
            req.add_header('User-Agent', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36')
            
            with urllib.request.urlopen(req) as response:
                if response.status == 200:
                    channel_data = json.loads(response.read().decode('utf-8'))
                    
                    if 'content' in channel_data and channel_data['content']:
                        is_live = channel_data['content'].get('openLive', False)
                        channel_name = channel_data['content'].get('channelName', '알 수 없음')
                        
                        logger.info(f"채널명: {channel_name}")
                        logger.info(f"방송 상태: {'방송 중' if is_live else '방송 종료'}")
                        
                        # 방송이 꺼져있어도 채팅방 접근 허용 (Old version 방식)
                        if not is_live:
                            logger.warning("현재 방송이 꺼져있지만 채팅방 연결을 시도합니다.")
                    else:
                        logger.error("채널 정보를 찾을 수 없습니다.")
                        return False
                else:
                    logger.error(f"채널 정보 조회 실패: {response.status}")
                    return False
        except Exception as e:
            logger.error(f"채널 정보 조회 실패: {e}")
            return False
        
        # 2단계: 라이브 상세 정보에서 채팅 채널 ID 획득
        api_endpoints = [
            f"https://api.chzzk.naver.com/service/v2/channels/{self.channel_id}/live-detail",
            f"https://api.chzzk.naver.com/service/v1/channels/{self.channel_id}/live-detail",
            f"https://api.chzzk.naver.com/service/v1/channels/{self.channel_id}/live-status",
        ]
        
        for endpoint in api_endpoints:
            try:
                req = urllib.request.Request(endpoint)
                req.add_header('User-Agent', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36')
                
                with urllib.request.urlopen(req) as response:
                    if response.status == 200:
                        data = json.loads(response.read().decode('utf-8'))
                        
                        # 채팅 채널 ID 추출 (Old version과 동일한 방식)
                        chat_channel_id = None
                        if 'content' in data and data['content'] and 'chatChannelId' in data['content']:
                            chat_channel_id = data['content']['chatChannelId']
                        elif 'content' in data and data['content'] and 'live' in data['content'] and data['content']['live'] and 'chatChannelId' in data['content']['live']:
                            chat_channel_id = data['content']['live']['chatChannelId']
                        
                        if chat_channel_id:
                            self.chat_channel_id = chat_channel_id
                            logger.info(f"채팅 채널 ID 획득: {self.chat_channel_id}")
                            return True
                        else:
                            logger.debug(f"채팅 채널 ID 없음 - 다음 API 시도: {endpoint}")
                            continue
                    else:
                        logger.debug(f"API 응답 실패 ({response.status}) - 다음 API 시도: {endpoint}")
                        continue
            except Exception as e:
                logger.debug(f"API 호출 실패 - 다음 API 시도: {endpoint}, 오류: {e}")
                continue
        
        # 최후의 수단: 채널 ID를 채팅 채널 ID로 사용
        logger.info("채널 ID를 채팅 채널 ID로 사용합니다.")
        self.chat_channel_id = self.channel_id
        return True
    
    def get_access_token_sync(self):
        """액세스 토큰 획득 - 성공했던 간단한 로직"""
        if not self.chat_channel_id:
            logger.error("채팅 채널 ID가 없습니다.")
            return False
            
        url = f"https://comm-api.game.naver.com/nng_main/v1/chats/access-token?channelId={self.chat_channel_id}&chatType=STREAMING"
        
        try:
            req = urllib.request.Request(url)
            req.add_header('User-Agent', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36')
            
            with urllib.request.urlopen(req) as response:
                if response.status == 200:
                    data = json.loads(response.read().decode('utf-8'))
                    
                    if 'content' in data and 'accessToken' in data['content']:
                        self.access_token = data['content']['accessToken']
                        logger.info("액세스 토큰 획득 성공")
                        return True
                    else:
                        logger.error("액세스 토큰을 찾을 수 없습니다.")
                        return False
                else:
                    logger.error(f"액세스 토큰 조회 실패: {response.status}")
                    return False
        except Exception as e:
            logger.error(f"액세스 토큰 조회 실패: {e}")
            return False
    
    async def connect(self):
        """웹소켓 연결 - Old version과 동일한 검증된 로직"""
        logger.info("=== 채팅방 연결 시작 ===")
        
        # 1단계: 채팅 채널 ID 획득
        logger.info("1단계: 채팅 채널 ID 획득 중...")
        loop = asyncio.get_event_loop()
        if not await loop.run_in_executor(None, self.get_chat_channel_id_sync):
            logger.error("❌ 채팅 채널 ID 획득 실패")
            return False
        logger.info("✅ 채팅 채널 ID 획득 성공")
        
        # 2단계: 액세스 토큰 획득
        logger.info("2단계: 액세스 토큰 획득 중...")
        if not await loop.run_in_executor(None, self.get_access_token_sync):
            logger.error("❌ 액세스 토큰 획득 실패")
            return False
        logger.info("✅ 액세스 토큰 획득 성공")
        
        # 3단계: 웹소켓 연결
        logger.info("3단계: 웹소켓 연결 시도 중...")
        for i, endpoint in enumerate(self.endpoints, 1):
            try:
                logger.info(f"웹소켓 서버 {i}/{len(self.endpoints)} 연결 시도: {endpoint}")
                headers = {
                    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
                    "Origin": "https://chzzk.naver.com"
                }
                self.websocket = await websockets.connect(endpoint, additional_headers=headers)
                self.is_connected = True
                logger.info(f"✅ 웹소켓 연결 성공: {endpoint}")
                logger.info("=== 채팅방 연결 완료 ===")
                return True
            except Exception as e:
                logger.warning(f"❌ 웹소켓 서버 {i} 연결 실패: {e}")
                continue
        
        logger.error("❌ 모든 웹소켓 서버 연결 실패")
        logger.error("💡 해결 방법:")
        logger.error("   1. 네트워크 연결 확인")
        logger.error("   2. 채널 ID가 올바른지 확인")
        logger.error("   3. 방화벽 설정 확인")
        return False
    
    async def send_join_message(self):
        """채팅방 참가 메시지 전송 - 성공했던 간단한 로직"""
        if not self.chat_channel_id or not self.access_token:
            logger.error("채팅 채널 ID 또는 액세스 토큰이 없습니다.")
            return
        
        connect_message = {
            "ver": "2",
            "cmd": 100,
            "svcid": "game",
            "cid": self.chat_channel_id,
            "bdy": {
                "uid": None,
                "devType": 2001,
                "accTkn": self.access_token,
                "auth": "READ"
            },
            "tid": 1
        }
        
        try:
            await self.websocket.send(json.dumps(connect_message))
            logger.info("채팅방 참가 메시지 전송 완료")
        except Exception as e:
            logger.error(f"참가 메시지 전송 실패: {e}")
    
    async def listen_messages(self, message_callback=None):
        """메시지 수신 대기 - 연결 안정성 개선"""
        logger.info("채팅 메시지 수신 시작")
        
        # 하트비트 태스크 시작
        heartbeat_task = asyncio.create_task(self.keep_alive())
        
        try:
            message_count = 0
            async for message in self.websocket:
                try:
                    if message.strip():
                        data = json.loads(message)
                        await self.handle_message(data, message_callback)
                        message_count += 1
                        
                        # 주기적으로 연결 상태 확인
                        if message_count % 100 == 0:
                            logger.debug(f"메시지 {message_count}개 처리됨 - 연결 상태 양호")
                            
                except json.JSONDecodeError:
                    logger.debug(f"JSON 파싱 실패: {message[:100] if len(message) > 100 else message}...")
                except Exception as e:
                    logger.error(f"메시지 처리 오류: {e}")
                    # 메시지 처리 오류는 연결을 끊지 않고 계속 진행
                    continue
                    
        except websockets.exceptions.ConnectionClosed as e:
            logger.warning(f"웹소켓 연결 종료: {e}")
            self.is_connected = False
        except websockets.exceptions.InvalidState as e:
            logger.warning(f"웹소켓 상태 오류 - 재연결 필요: {e}")
            self.is_connected = False
        except websockets.exceptions.WebSocketException as e:
            logger.error(f"웹소켓 오류: {e}")
            self.is_connected = False
        except Exception as e:
            logger.error(f"예상치 못한 메시지 수신 오류: {e}")
            self.is_connected = False
        finally:
            logger.info(f"메시지 수신 종료됨 (총 {message_count}개 처리)")
            heartbeat_task.cancel()
            try:
                await heartbeat_task
            except asyncio.CancelledError:
                pass
    
    async def handle_message(self, data, message_callback=None):
        """메시지 처리"""
        if isinstance(data, dict):
            # 하트비트 응답
            if data.get('cmd') == 0 and data.get('ver') == '2':
                heartbeat_response = {"ver": "2", "cmd": 10000}
                try:
                    await self.websocket.send(json.dumps(heartbeat_response))
                except Exception as e:
                    logger.error(f"하트비트 응답 실패: {e}")
                return
            
            # 연결 응답 처리
            elif data.get('cmd') == 10000:
                ret_code = data.get('retCode')
                if ret_code == 200:
                    logger.info("채팅방 연결 성공!")
                else:
                    logger.error(f"연결 실패: {data.get('retMsg', '')}")
                return
            
            # 채팅 메시지 처리
            elif 'bdy' in data:
                await self.process_chat_message(data, message_callback)
    
    async def process_chat_message(self, data, message_callback=None):
        """채팅 메시지 파싱 및 저장 - 중복 방지 및 안정성 개선"""
        try:
            cmd = data.get('cmd')
            bdy_list = data.get('bdy', [])
            
            if not bdy_list:
                return
            
            # 배열의 모든 메시지 처리 (Old version은 단일 메시지만 처리했지만 여기서는 개선)
            if isinstance(bdy_list, list):
                # 여러 메시지가 한번에 올 수 있으므로 모두 처리
                for bdy in bdy_list:
                    await self._process_single_message(bdy, cmd, message_callback)
            else:
                # 단일 메시지 처리
                await self._process_single_message(bdy_list, cmd, message_callback)
                    
        except Exception as e:
            logger.error(f"채팅 메시지 파싱 오류: {e}")
    
    async def _process_single_message(self, bdy, cmd, message_callback=None):
        """단일 메시지 처리 - 중복 방지 로직 포함"""
        try:
            # 메시지 ID 기반 중복 체크
            message_id = bdy.get('msgId') or bdy.get('id') or bdy.get('uid', '') + str(bdy.get('msgTime', ''))
            
            # 빈 메시지나 중복 메시지 필터링
            message_text = bdy.get('msg', '').strip()
            if not message_text or not message_id:
                return
            
            # 프로필 정보 파싱
            profile_str = bdy.get('profile', '{}')
            try:
                profile = json.loads(profile_str) if isinstance(profile_str, str) else profile_str
            except json.JSONDecodeError:
                profile = {}
            
            # 사용자 역할 확인
            user_role = profile.get('userRoleCode', 'common_user')
            is_streamer = (user_role == 'streamer')
            
            # 후원 메시지 확인 (추정)
            is_donation = cmd == 93102 or bdy.get('payAmount', 0) > 0
            
            # 배지 및 제목 정보
            badge = profile.get('badge', {})
            title = profile.get('title', {})
            
            chat_data = {
                'id': message_id,  # 중복 방지용 ID 추가
                'type': 'donation' if is_donation else 'chat',
                'timestamp': datetime.now().strftime('%H:%M:%S'),
                'user_id': bdy.get('uid', ''),
                'nickname': profile.get('nickname', '익명'),
                'message': message_text,
                'is_streamer': is_streamer,
                'user_role': user_role,
                'badge_url': badge.get('imageUrl', '') if badge else '',
                'title_name': title.get('name', '') if title else '',
                'title_color': title.get('color', '#FFFFFF') if title else '#FFFFFF',
                'profile_image': profile.get('profileImageUrl', ''),
                'verified': profile.get('verifiedMark', False),
                'amount': bdy.get('payAmount', 0) if is_donation else 0,
                'msg_time': bdy.get('msgTime', 0)  # 메시지 시간 추가
            }
            
            # 콜백 함수가 있으면 호출
            if message_callback:
                message_callback(chat_data)
            
            # 콘솔 출력
            role_emoji = "👑" if is_streamer else ("💰" if is_donation else "💬")
            amount_text = f" ({chat_data['amount']}원)" if is_donation else ""
            print(f"{role_emoji} [{chat_data['timestamp']}] {chat_data['nickname']}: {chat_data['message']}{amount_text}")
                    
        except Exception as e:
            logger.error(f"단일 메시지 처리 오류: {e}")
    
    async def keep_alive(self):
        """하트비트 전송으로 연결 유지"""
        try:
            while self.is_connected:
                await asyncio.sleep(20)  # 20초마다
                if self.websocket and not self.websocket.closed:
                    heartbeat = {"ver": "2", "cmd": 0}
                    await self.websocket.send(json.dumps(heartbeat))
        except Exception as e:
            logger.debug(f"하트비트 종료: {e}")
    
    async def disconnect(self):
        """연결 종료"""
        if self.websocket:
            await self.websocket.close()
            self.is_connected = False 

File: src/chat-client.js
#!/usr/bin/env node

/**
 * CHZZK 채팅 클라이언트
 * 실시간 채팅 메시지를 수신하여 콘솔에 출력합니다.
 * 
 * @author ChzzkStreamDeck
 * @version 2.0.0
 */

let fetch;

/**
 * node-fetch ES Module을 동적으로 로드
 */
async function loadFetch() {
    if (!fetch) {
        fetch = (await import('node-fetch')).default;
    }
    return fetch;
}

const WebSocket = require('ws');

/**
 * CHZZK 채팅 클라이언트 클래스
 */
class ChzzkChatClient {
    constructor(channelId, options = {}) {
        this.channelId = channelId;
        this.chatChannelId = null;
        this.accessToken = null;
        this.websocket = null;
        this.isConnected = false;
        this.heartbeatInterval = null;
        
        // 옵션 설정
        this.options = {
            reconnectAttempts: 3,
            heartbeatInterval: 20000,
            connectionTimeout: 5000,
            verbose: false,
            ...options
        };
        
        // HTTP 요청 헤더
        this.headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Accept': 'application/json, text/plain, */*',
            'Accept-Language': 'ko-KR,ko;q=0.9,en;q=0.8',
            'Referer': 'https://chzzk.naver.com/',
            'Origin': 'https://chzzk.naver.com'
        };
    }

    /**
     * 클라이언트 시작
     */
    async start() {
        try {
            this.log('🚀 CHZZK 채팅 클라이언트 시작');
            this.log(`📺 채널 ID: ${this.channelId}`);
            
            // API 호출 단계별 실행
            await this.initializeConnection();
            await this.connectWebSocket();
            
            this.log('✅ 채팅 클라이언트 초기화 완료');
            
        } catch (error) {
            this.error(`❌ 클라이언트 시작 실패: ${error.message}`);
            throw error;
        }
    }

    /**
     * 연결 초기화 (API 호출)
     */
    async initializeConnection() {
        // 1. 채널 정보 가져오기
        const channelInfo = await this.getChannelInfo();
        if (!channelInfo) {
            throw new Error('채널 정보를 가져올 수 없습니다.');
        }
        
        // 2. 라이브 상태 확인
        const liveStatus = await this.getLiveStatus();
        if (!liveStatus) {
            throw new Error('라이브 상태를 확인할 수 없습니다.');
        }
        
        // 3. 액세스 토큰 가져오기
        const accessToken = await this.getAccessToken();
        if (!accessToken) {
            throw new Error('액세스 토큰을 가져올 수 없습니다.');
        }
    }

    /**
     * 채널 기본 정보 조회
     */
    async getChannelInfo() {
        try {
            const fetchFunction = await loadFetch();
            const response = await fetchFunction(
                `https://api.chzzk.naver.com/service/v1/channels/${this.channelId}`, 
                { headers: this.headers }
            );
            
            if (response.ok) {
                const data = await response.json();
                if (data.code === 200 && data.content) {
                    this.log(`✅ 채널 정보: ${data.content.channelName || 'N/A'}`);
                    return data.content;
                }
            }
            
            this.warn('⚠️ 채널 정보 요청 실패');
            return null;
            
        } catch (error) {
            this.warn(`⚠️ 채널 정보 오류: ${error.message}`);
            return null;
        }
    }

    /**
     * 라이브 상태 조회
     */
    async getLiveStatus() {
        try {
            const fetchFunction = await loadFetch();
            const response = await fetchFunction(
                `https://api.chzzk.naver.com/polling/v2/channels/${this.channelId}/live-status`, 
                { headers: this.headers }
            );
            
            if (response.ok) {
                const data = await response.json();
                if (data.code === 200 && data.content) {
                    const content = data.content;
                    this.chatChannelId = content.chatChannelId;
                    this.log(`✅ 라이브 상태: ${content.status || content.liveStatus}`);
                    return content;
                }
            }
            
            this.warn('⚠️ 라이브 상태 확인 실패');
            return null;
            
        } catch (error) {
            this.warn(`⚠️ 라이브 상태 오류: ${error.message}`);
            return null;
        }
    }

    /**
     * 채팅 액세스 토큰 조회
     */
    async getAccessToken() {
        if (!this.chatChannelId) {
            this.error('❌ 채팅 채널 ID가 없습니다.');
            return null;
        }

        try {
            const fetchFunction = await loadFetch();
            const response = await fetchFunction(
                `https://comm-api.game.naver.com/nng_main/v1/chats/access-token?channelId=${this.chatChannelId}&chatType=STREAMING`, 
                { headers: this.headers }
            );
            
            if (response.ok) {
                const data = await response.json();
                if (data.code === 200 && data.content) {
                    this.accessToken = data.content.accessToken;
                    this.log('✅ 액세스 토큰 획득 완료');
                    return data.content;
                }
            }
            
            this.warn('⚠️ 액세스 토큰 요청 실패');
            return null;
            
        } catch (error) {
            this.warn(`⚠️ 액세스 토큰 오류: ${error.message}`);
            return null;
        }
    }

    /**
     * WebSocket 연결
     */
    async connectWebSocket() {
        this.log('🔗 WebSocket 연결 시도...');
        
        if (!this.accessToken || !this.chatChannelId) {
            throw new Error('연결에 필요한 정보가 부족합니다.');
        }

        // kr-ss1 ~ kr-ss10 서버 순차 시도
        for (let serverNum = 1; serverNum <= 10; serverNum++) {
            const wsUrl = `wss://kr-ss${serverNum}.chat.naver.com/chat?channelId=${this.chatChannelId}&accessToken=${this.accessToken}`;
            
            try {
                this.verbose(`🔗 WebSocket 서버 시도 ${serverNum}/10: kr-ss${serverNum}`);
                
                if (await this.tryConnectToServer(wsUrl, serverNum)) {
                    return; // 연결 성공 시 종료
                }
                
            } catch (error) {
                this.verbose(`❌ kr-ss${serverNum} 연결 실패: ${error.message}`);
                
                // 다음 서버 시도 전 정리
                this.cleanupWebSocket();
                
                if (serverNum < 10) {
                    await this.sleep(100);
                }
            }
        }
        
        throw new Error('모든 WebSocket 서버 연결 실패');
    }

    /**
     * 특정 서버로 WebSocket 연결 시도
     */
    async tryConnectToServer(wsUrl, serverNum) {
        return new Promise((resolve, reject) => {
            this.websocket = new WebSocket(wsUrl);
            
            const timeout = setTimeout(() => {
                reject(new Error('연결 타임아웃'));
            }, this.options.connectionTimeout);
            
            this.websocket.on('open', () => {
                clearTimeout(timeout);
                this.log(`✅ WebSocket 연결 성공: kr-ss${serverNum}`);
                this.isConnected = true;
                
                this.setupWebSocketHandlers();
                this.authenticateChat();
                this.startHeartbeat();
                
                resolve(true);
            });
            
            this.websocket.on('error', (error) => {
                clearTimeout(timeout);
                reject(error);
            });
        });
    }

    /**
     * WebSocket 이벤트 핸들러 설정
     */
    setupWebSocketHandlers() {
        this.websocket.on('message', (data) => {
            try {
                const message = JSON.parse(data.toString());
                this.handleMessage(message);
            } catch (error) {
                this.warn(`⚠️ 메시지 파싱 실패: ${error.message}`);
            }
        });
        
        this.websocket.on('error', (error) => {
            this.error(`❌ WebSocket 오류: ${error.message}`);
        });
        
        this.websocket.on('close', (code, reason) => {
            if (this.isConnected) {
                this.log(`🔌 WebSocket 연결 종료: ${code} - ${reason}`);
                this.isConnected = false;
                this.stopHeartbeat();
            }
        });
    }

    /**
     * 채팅 인증
     */
    authenticateChat() {
        const authMessage = {
            ver: "2",
            cmd: 100,
            svcid: "game",
            cid: this.chatChannelId,
            bdy: {
                uid: null,
                devType: 2001,
                accTkn: this.accessToken,
                auth: "READ"
            },
            tid: 1
        };
        
        this.websocket.send(JSON.stringify(authMessage));
        this.log('🔓 채팅 인증 요청 전송');
    }

    /**
     * 하트비트 시작
     */
    startHeartbeat() {
        this.heartbeatInterval = setInterval(() => {
            if (this.websocket && this.websocket.readyState === WebSocket.OPEN) {
                const heartbeatMessage = { ver: "2", cmd: 0 };
                this.websocket.send(JSON.stringify(heartbeatMessage));
                this.verbose('💓 하트비트 전송');
            }
        }, this.options.heartbeatInterval);
    }

    /**
     * 하트비트 중지
     */
    stopHeartbeat() {
        if (this.heartbeatInterval) {
            clearInterval(this.heartbeatInterval);
            this.heartbeatInterval = null;
        }
    }

    /**
     * WebSocket 정리
     */
    cleanupWebSocket() {
        if (this.websocket) {
            this.websocket.removeAllListeners();
            if (this.websocket.readyState === WebSocket.OPEN || 
                this.websocket.readyState === WebSocket.CONNECTING) {
                this.websocket.close();
            }
            this.websocket = null;
        }
    }

    /**
     * 메시지 처리
     */
    handleMessage(message) {
        switch (message.cmd) {
            case 0:
                // 서버 하트비트 요청 - 응답 필요
                const response = { ver: "2", cmd: 10000 };
                this.websocket.send(JSON.stringify(response));

                break;
                
            case 10100:
                // 인증 완료
                this.log('💬 채팅 연결 완료');
                break;
                
            case 93101:
                // 채팅 메시지
                this.handleChatMessage(message);
                break;
                
            default:

        }
    }

    /**
     * 채팅 메시지 처리
     */
    handleChatMessage(message) {
        if (!message.bdy || !Array.isArray(message.bdy) || message.bdy.length === 0) {
            return;
        }

        for (const chatData of message.bdy) {
            try {
                const nickname = this.extractNickname(chatData);
                const content = chatData.msg || chatData.content || '';
                
                if (content.trim()) {
                    // 이모티콘 정보 추출
                    const emoticons = this.extractEmoticons(chatData);
                    
                    if (emoticons && Object.keys(emoticons).length > 0) {
                        // 이모티콘이 있는 경우 JSON 형태로 출력
                        const messageData = {
                            username: nickname,
                            message: content,
                            extras: { emojis: emoticons }
                        };
                        console.log(`CHAT_JSON:${JSON.stringify(messageData)}`);
                    } else {
                        // 기존 형태로 출력
                        console.log(`[${nickname}]: ${content}`);
                    }
                }
                
            } catch (error) {

            }
        }
    }

    /**
     * 이모티콘 정보 추출
     */
    extractEmoticons(chatData) {
        try {
            if (chatData.extras) {
                let extras = {};
                
                if (typeof chatData.extras === 'string') {
                    extras = JSON.parse(chatData.extras);
                } else if (typeof chatData.extras === 'object') {
                    extras = chatData.extras;
                }
                
                if (extras.emojis && typeof extras.emojis === 'object') {
                    return extras.emojis;
                }
            }
            
            return null;
            
        } catch (error) {

            return null;
        }
    }

    /**
     * 닉네임 추출
     */
    extractNickname(chatData) {
        try {
            let profile = {};
            
            if (typeof chatData.profile === 'string') {
                profile = JSON.parse(chatData.profile);
            } else if (typeof chatData.profile === 'object') {
                profile = chatData.profile || {};
            }
            
            return profile.nickname || '익명';
            
        } catch (error) {
            return '익명';
        }
    }

    /**
     * 연결 종료
     */
    disconnect() {
        this.stopHeartbeat();
        this.cleanupWebSocket();
        this.isConnected = false;
    }

    /**
     * 상태 조회
     */
    getStatus() {
        return {
            connected: this.isConnected,
            channelId: this.channelId,
            chatChannelId: this.chatChannelId,
            hasAccessToken: !!this.accessToken
        };
    }

    // 로깅 메서드들
    log(message) { console.log(message); }
    error(message) { console.error(message); }
    warn(message) { console.log(message); }
    verbose(message) { if (this.options.verbose) console.log(message); }
    
    // 유틸리티 메서드
    sleep(ms) { return new Promise(resolve => setTimeout(resolve, ms)); }
}

// 명령행에서 직접 실행하는 경우
if (require.main === module) {
    const channelId = process.argv[2];
    const verbose = process.argv.includes('--verbose');
    
    if (!channelId) {
        console.error('❌ 채널 ID가 필요합니다.');
        console.error('사용법: node src/chat-client.js <channelId> [--verbose]');
        process.exit(1);
    }
    
    const client = new ChzzkChatClient(channelId, { verbose });
    
    // 프로세스 종료 시 정리
    process.on('SIGINT', () => {
        client.disconnect();
        process.exit(0);
    });
    
    process.on('SIGTERM', () => {
        client.disconnect();
        process.exit(0);
    });
    
    // 클라이언트 시작
    client.start().catch(error => {
        console.error(`❌ 시작 실패: ${error.message}`);
        process.exit(1);
    });
}

module.exports = ChzzkChatClient; 